### PR TITLE
`gen-attributions` CLI tool

### PR DESCRIPTION
**Pull request by A-Hilaly**
Introduces a new tool that helps ACK developers and Go users in
generating software attributions file (known as `ATTRIBUTIONS.md` in ACK
land).

`gen-attributions` parses a given `go.mod` file and generates its
corresponding attributions file. All collected information (`go.mod`
files and the licenses) are obtained (by default) from the official Go
proxy `proxy.golang.org`.

The easiest way to use `gen-attributions` is to run it inside the directory
of your Go project.

```bash
gen-attributions --debug # default generated file name is ATTRIBUTIONS.md
```

You can also set other parameters like the graph max depth, input/output
and the template files used to generate the software attributions file.

```bash
gen-attributions --depth 5 --modfile go.mod --output ATTRIBUTIONS.md

gen-attributions \
    --attr-header-template $(HEADER_TMP)\
    --attr-block-template $(BLOCK_TMP)
```

You can also use `gen-attributions` CLI to visualize the dependencies
graph.

```bash
gen-attributions --show-graph

gen-attributions
└── golang.org/x/mod@v0.5.1 BSD-3-Clause
    ├── golang.org/x/crypto@v0.0.0-20191011191535-87dc89f01550 BSD-3-Clause
    │   ├── golang.org/x/net@v0.0.0-20190404232315-eb5bcb51f2a3 BSD-3-Clause
    │   │   ├── golang.org/x/crypto@v0.0.0-20190308221718-c2843e01d9a2 BSD-3-Clause
    │   │   │   └── golang.org/x/sys@v0.0.0-20190215142949-d0b11bdaac8a BSD-3-Clause
    │   │   └── golang.org/x/text@v0.3.0 BSD-3-Clause
    │   └── golang.org/x/sys@v0.0.0-20190412213103-97732733099d BSD-3-Clause
    ├── golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e BSD-3-Clause
    │   ├── golang.org/x/net@v0.0.0-20190620200207-3b0461eec859 BSD-3-Clause
    │   │   ├── golang.org/x/crypto@v0.0.0-20190308221718-c2843e01d9a2 BSD-3-Clause
    │   │   │   └── golang.org/x/sys@v0.0.0-20190215142949-d0b11bdaac8a BSD-3-Clause
    │   │   ├── golang.org/x/sys@v0.0.0-20190215142949-d0b11bdaac8a BSD-3-Clause
    │   │   └── golang.org/x/text@v0.3.0 BSD-3-Clause
    │   ├── golang.org/x/sync@v0.0.0-20190423024810-112230192c58 BSD-3-Clause
    │   └── golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7 BSD-3-Clause
    └── golang.org/x/xerrors@v0.0.0-20191011141410-1b5146add898 BSD-3-Clause
```

To debug the process of dependencies graph exploration use the `--debug`
flag

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

### CODE REVIEW

File: tools/gen-attributions/ATTRIBUTIONS.md [hunk 1]
```diff
@@ -0,0 +1,1251 @@
...
+
+ACK controllers depends on a number of Open Source Go packages. Direct 
+dependencies are listed in the ./go.mod file.
+Those direct package dependencies have some dependencies of their own (known as
+"transitive dependencies")
+
+In this part of the Attribution document, we list our dependent packages and
+include an indication of the Open Source License under which that package is
+distributed. For any package *NOT* distributed under the terms of the Apache
+License version 2.0, we include the full text of the package's License below.
```
Comment (jaypipes) [inline tools/gen-attributions/ATTRIBUTIONS.md:21]: Below, I see that the spf13/cobra dependency actually *DOES* include the full text of the Apache 2.0 license... is that intended?

File: tools/gen-attributions/cmd/render.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/mod/module"
+)
+
+const (
+	defaultHeader = `# Open Source Software Attribution
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/render.go:27]: did you want to place this in the `templates.go` file instead?

**Review by jaypipes — CHANGES_REQUESTED** [review]
@A-Hilaly man, this is friggin cool. Love it! :) Just a couple tiny things to address inline.

File: tools/gen-attributions/ATTRIBUTIONS.md (hunk 1 above)
Comment (A-Hilaly) [reply tools/gen-attributions/ATTRIBUTIONS.md:21]: Initially, i wanted to add that feature but completely forgot. Sending a commit for that very soon.

Comment (A-Hilaly) [reply tools/gen-attributions/ATTRIBUTIONS.md:21]: Addressed in 91d6de1363d489acc0d22edc12e9765ebcccc93f, this reduced the size of the ATTRIBUTIONS.md by 900 lines... foewww

File: tools/gen-attributions/cmd/goproxy.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		return nil, err
+	}
+	if resp.StatusCode == 200 {
+		return body, nil
+	}
+	if resp.StatusCode == 410 && strings.Contains(string(body), "invalid escaped module path") {
+		return nil, ErrInvalidEscapedModulePath
+	}
+
+	return nil, fmt.Errorf("unhanled response status code: %v", resp.StatusCode)
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/goproxy.go:55]:
```suggestion
	return nil, fmt.Errorf("unhandled response status code: %v", resp.StatusCode)
```

File: tools/gen-attributions/cmd/license.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+const (
+	LicenseUnknown LicenseType = iota
+	LicenseApache20
+	LicenseMIT
+	LicenseBSD3Clause
+	LicenseISC
+	LicenseOther
+
+	defaultConfidenceTreshHold = 0.9
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/license.go:23]:
```suggestion
	defaultConfidenceTreshold = 0.9
```

File: tools/gen-attributions/cmd/license.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	default:
+		return LicenseOther
+	}
+}
+
+// License represents a sotfware distribution and usage license
+type License struct {
+	// The name of the license
+	Name string
+	// If the name of the license if unkown, this field will be
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/license.go:47]:
```suggestion
	// If the name of the license is unknown, this field will be
```

File: tools/gen-attributions/cmd/license.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	// filled
+	Data []byte
+}
+
+// newLicenseClassifier instantiate a new classifer and returns
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/license.go:52]:
```suggestion
// newLicenseClassifier instantiates a new classifier and returns
```

File: tools/gen-attributions/cmd/license.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"errors"
+
+	"github.com/google/licenseclassifier"
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/license.go:6]: cool. :)

File: tools/gen-attributions/cmd/license.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// licenseClassifierWrapper is a wrapper around licenseclassifier.License
+type licenseClassifierWrapper struct {
+	classifier *licenseclassifier.License
+}
+
+// detectLicense takes the byte of a given license and returns its name
+func (lcw *licenseClassifierWrapper) detectLicense(data []byte) (string, error) {
+	matches := lcw.classifier.MultipleMatch(string(data), true)
```
Comment (jaypipes) [inline tools/gen-attributions/cmd/license.go:74]: You could probably just use [`Classifier.NearestMatch()`](https://github.com/google/licenseclassifier/blob/c913e304a1534c4580fa70c2c3af5cd85d99fc9c/classifier.go#L151) instead here.

**Review by jaypipes — APPROVED** [review]
This is fantastic, @A-Hilaly. Left a few typos to fix up and a couple suggestions but I'd be cool merging this as-is.

**Pull request closed without merging**